test/include/seqan3/test/sandboxed_path.hpp

```diff
@@ -0,0 +1,461 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2021, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2021, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/* \file
+ * \brief Provides the sandboxed_path and related free functions
+ * \author Simon Gene <simon.gottlieb AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/std/filesystem>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3::test
+{
+
+/** \brief Utility class to stay inside a sandbox path.
+ *
+ * sandboxed_path inherits from std::filesystem::path and behaves mostly
+ * like it. In addition, it receives a sandbox directory at construction time.
+ * Functions are overloaded and a check for the invariant is added.
+ *
+ *  Invariant:
+ *  - sandboxed_path is always converted to an absolute path
+ *  - sandboxed_path always points to a file inside a given sandbox directory.
+ *  - sandbox directory is immutable during the life cycle of a sandboxed_path
+ *
+ * Caveat:
+ *  - relative paths are not possible
+ *  - some functions will leave the sandboxed environment.
+ *    - e.g.: calling relative_path() leaves the environment of sandboxed_path.
```

Comment:
`relative_path()` takes a `std::filesystem::path` because of slicing it will accept a `sandboxed_path` and returns a `std::filesystem::path`. This returned `std::filesystem::path` is of course outside of the handle of `sandboxed_path`. No gurantees we can make. We could offer our own `relative_path()` function, but it doesn't really make sense, since one of the constraints of `sandboxed_path` is that we operate on absolute paths (to avoid other issues).
All in all it would take a lot more work to get a 100% tight solution while being a drop in replacement for std::filesystem::path and easy to use.